core/services/ccip/orm.go

```diff
@@ -0,0 +1,169 @@
+
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/assets"
+)
+
+type GasPrice struct {
+	SourceChainSelector uint64
+	GasPrice            *assets.Wei
+	CreatedAt           time.Time
+}
+
+type GasPriceUpdate struct {
+	SourceChainSelector uint64
+	GasPrice            *assets.Wei
+}
+
+type TokenPrice struct {
+	TokenAddr  string
+	TokenPrice *assets.Wei
+	CreatedAt  time.Time
+}
+
+type TokenPriceUpdate struct {
+	TokenAddr  string
+	TokenPrice *assets.Wei
+}
+
+//go:generate mockery --quiet --name ORM --output ./mocks/ --case=underscore
+type ORM interface {
+	GetGasPricesByDestChain(ctx context.Context, destChainSelector uint64) ([]GasPrice, error)
+	GetTokenPricesByDestChain(ctx context.Context, destChainSelector uint64) ([]TokenPrice, error)
+
+	InsertGasPricesForDestChain(ctx context.Context, destChainSelector uint64, jobId int32, gasPrices []GasPriceUpdate) error
+	InsertTokenPricesForDestChain(ctx context.Context, destChainSelector uint64, jobId int32, tokenPrices []TokenPriceUpdate) error
+
+	ClearGasPricesByDestChain(ctx context.Context, destChainSelector uint64, to time.Time) error
+	ClearTokenPricesByDestChain(ctx context.Context, destChainSelector uint64, to time.Time) error
+}
+
+type orm struct {
+	ds sqlutil.DataSource
+}
+
+var _ ORM = (*orm)(nil)
+
+const (
+	gasTableName   = "ccip.observed_gas_prices"
+	tokenTableName = "ccip.observed_token_prices"
+)
+
+func NewORM(ds sqlutil.DataSource) (ORM, error) {
+	if ds == nil {
+		return nil, fmt.Errorf("datasource to CCIP NewORM cannot be nil")
+	}
+
+	return &orm{
+		ds: ds,
+	}, nil
+}
+
+func (o *orm) GetGasPricesByDestChain(ctx context.Context, destChainSelector uint64) ([]GasPrice, error) {
+	var gasPrices []GasPrice
+	stmt := fmt.Sprintf(`
+		SELECT DISTINCT ON (source_chain_selector)
+		source_chain_selector, gas_price, created_at
+		FROM %s
+		WHERE chain_selector = $1
+		ORDER BY source_chain_selector, created_at DESC;
+	`, gasTableName)
+	err := o.ds.SelectContext(ctx, &gasPrices, stmt, destChainSelector)
+	if err != nil {
+		return nil, err
+	}
+
+	return gasPrices, nil
+}
+
+func (o *orm) GetTokenPricesByDestChain(ctx context.Context, destChainSelector uint64) ([]TokenPrice, error) {
+	var tokenPrices []TokenPrice
+	stmt := fmt.Sprintf(`
+		SELECT DISTINCT ON (token_addr)
+		token_addr, token_price, created_at
+		FROM %s
+		WHERE chain_selector = $1
+		ORDER BY token_addr, created_at DESC;
+
+	`, tokenTableName)
+	err := o.ds.SelectContext(ctx, &tokenPrices, stmt, destChainSelector)
+	if err != nil {
+		return nil, err
+	}
+
+	return tokenPrices, nil
+}
+
+func (o *orm) InsertGasPricesForDestChain(ctx context.Context, destChainSelector uint64, jobId int32, gasPrices []GasPriceUpdate) error {
+	if len(gasPrices) == 0 {
+		return nil
+	}
+
+	now := time.Now()
+	sqlStr := ""
+	var values []interface{}
+	for i, price := range gasPrices {
+		sqlStr += fmt.Sprintf("($%d,$%d,$%d,$%d,$%d),", i*5+1, i*5+2, i*5+3, i*5+4, i*5+5)
```

Comment:
There must be a way to avoid this manual construction. Do we need to use a Named query or something?

Reply:
you are right, found NamedExec, didn't know about it before